### PR TITLE
slogSender using positions in file instead of mmap

### PR DESCRIPTION
**Pull request by turadg**
refs: #8636

## Description

On the path to Node 20 compatibility, we need to drop the `bufferfromfile` dependency.  (It doesn't work in Node 20 and fixing it is prohibitive: https://github.com/agoric-labs/BufferFromFile/pull/2). We decided to give on on mmap and just use file offsets.

This refactors `flight-recorder.js` to abstract the file IO and implements an alternate IO using sync FS methods.

Then it changes the code using the mmap-ed IO to use the fs IO.

Finally it removes the mmap-ed version. The factoring could be simplified more after that but imo it helps to have the separate concerns in separate functions.

### Security Considerations

n/a

### Scaling Considerations

Writes are async. Clients can `forceFlush` when it matters. The tests do this to ensure they're reading after the writes.

### Documentation Considerations

Nothing relevant.

### Testing Considerations

I think the existing test coverage is enough.

### Upgrade Considerations

n/a

### CODE REVIEW

File: packages/telemetry/src/flight-recorder.js [hunk 1]
```diff
@@ -206,31 +181,212 @@ export const makeMemoryMappedCircularBuffer = async ({
       );
     }
 
+    writeRecord(record, firstWriteLength, circEnd);
+
+    header.setBigUint64(
```
Comment (michaelfig) [inline packages/telemetry/src/flight-recorder.js:170]: Where is this flushed to disk?

File: packages/telemetry/src/flight-recorder.js [hunk 2]
```diff
@@ -206,31 +181,212 @@ export const makeMemoryMappedCircularBuffer = async ({
...
+  const newArenaSize = await initializeCircularBuffer(
+    filename,
+    circularBufferSize,
+  );
+
+  /**
+   * @type {Uint8Array}
+   * BufferFromFile mmap()s the file into the process address space.
+   */
+  const fileBuf = BufferFromFile(filename).Uint8Array();
```
Comment (michaelfig) [inline packages/telemetry/src/flight-recorder.js]: This function still uses BufferFromFile.  Shouldn't the need for Node 20 compatibility require removing `makeMemoryMappedCircularBuffer` and all references to it, including `packages/telemetry/src/frcat-entrypoint.js` and `packages/telemetry/test/test-flight-recorder.js`?

File: packages/telemetry/src/flight-recorder.js [hunk 3]
```diff
@@ -206,31 +181,212 @@ export const makeMemoryMappedCircularBuffer = async ({
...
+      // Write to the beginning of the arena.
+      fs.writeSync(
+        file.fd,
+        record.subarray(firstWriteLength, record.byteLength),
+        undefined,
+        undefined,
+        header.byteLength,
+      );
+    }
+  };
```
Comment (michaelfig) [inline packages/telemetry/src/flight-recorder.js]: Something like?

```suggestion
     // Write out the updated file header.
     fs.writeSync(
       file.fd,
       headerBuffer,
     );
   };
```

**Review by michaelfig — COMMENTED** [review]
Approach and architecture looks good!

I don't know what your plan is for what you want to do before you take this PR out of draft, so please forgive me if I'm pointing out what you consider to be already obvious.

File: packages/telemetry/src/flight-recorder.js (hunk 2 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js]: yeah. I suppose I didn't explain well enough in the PR description: I thought it best to keep that one around until we're fully confident in the new one. It's trivial to remove later. 

this is why I made the issue "ref" not "close" the compatibility issue. I'll expand the description

File: packages/telemetry/src/flight-recorder.js (hunk 3 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js]: ah, thanks. I had something in an earlier iteration and forgot to come back to it. Forthcoming.

I was planning to have a callback for updating the value, but I see now that would write every time through the "overlap > 0n" loop. 

Yours is a good solution imo. I'll add it.

**Comment by turadg** [comment]
> I don't know what your plan is for what you want to do before you take this PR out of draft, so please forgive me if I'm pointing out what you consider to be already obvious.

To get it out of draft I want:
- CI passing
- some input on whether it's okay to use synchronous IO everywhere
- input on a test plan. I've never used this tool so I don't know all its uses and whether CI has enough coverage

**Comment by mhofman** [comment]
> We discussed using async IO but `slogSender` itself is synchronous so I don't see a safe way to make it async without changing the slogSender API. If that's advised I'll do that as well.

haven't reviewed or read any comment yet, but while the send is sync, there is an async flush operation which can be used to guarantee previous writes succeeded.

File: packages/telemetry/src/make-slog-sender.js [hunk 4]
```diff
@@ -18,8 +18,10 @@ export const DEFAULT_SLOGSENDER_AGENT = 'self';
  */
 const filterTruthy = arr => /** @type {any[]} */ (arr.filter(Boolean));
 
+/** @typedef {import('./index.js').MakeSlogSenderCommonOptions} MakeSlogSenderCommonOptions */
```
Comment (mhofman) [inline packages/telemetry/src/make-slog-sender.js]: Why the alias?

File: packages/telemetry/src/flight-recorder.js [hunk 5]
```diff
@@ -225,14 +225,25 @@ export const makeMemoryMappedCircularBuffer = async ({
...
+  return { readCircBuf, writeCircBuf };
+};
+
+/**
+ *
+ * @param {Pick<Awaited<ReturnType<typeof makeMemoryMappedCircularBuffer>>, 'writeCircBuf'>} circBuf
+ */
+export const makeSlogSenderFromBuffer = ({ writeCircBuf }) => {
+  const writeJSON = obj => {
+    const jsonObj = serializeSlogObj(obj);
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js]: This is a perf regression. The original function accepts 2 arguments: the `makeSlogSender` router pre-serializes the JS object if any slog sender says it supports a serialized object (through `usesJsonObject`). (I agree the naming is not great)

File: packages/telemetry/src/flight-recorder.js [hunk 6]
```diff
@@ -13,9 +13,8 @@
 // flush it to disk any particular time, but knows when reads happen, so there's
 // no coherency problem, and the speed is unaffected by disk write speeds.
 
-import BufferFromFile from 'bufferfromfile';
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js:15]: From what I understand of the commit layering, this commit should not remove this import

File: packages/telemetry/src/flight-recorder.js [hunk 7]
```diff
@@ -31,10 +31,14 @@ const I_ARENA_START = 4 * BigUint64Array.BYTES_PER_ELEMENT;
 
 const RECORD_HEADER_SIZE = BigUint64Array.BYTES_PER_ELEMENT;
 
+/**
+ * Initializes a circular buffer with the given size, creating the buffer file if it doesn't exist or is not large enough.
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js:25]: FYI I've observed that with the current mmap implementation, resizing does not actually work (crashes if I remember, I forgot to write down my findings in https://github.com/Agoric/agoric-sdk/issues/8425)

File: packages/telemetry/src/flight-recorder.js
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js:1]: Nit: if possible, split the refactor/moving around of functions into a separate commit from adding the new implementation.

**Comment by turadg** [comment]
> > We discussed using async IO but `slogSender` itself is synchronous so I don't see a safe way to make it async without changing the slogSender API. If that's advised I'll do that as well.
> 
> haven't reviewed or read any comment yet, but while the send is sync, there is an async flush operation which can be used to guarantee previous writes succeeded.

Thanks for pointing that out. I'm convinced I can make the writes async without any API changes. But it does add a bit of work so please let me know whether you think it's necessary.

File: packages/telemetry/src/flight-recorder.js [hunk 8]
```diff
@@ -217,18 +181,186 @@ export const makeMemoryMappedCircularBuffer = async ({
...
+  const newArenaSize = await initializeCircularBuffer(
+    filename,
+    circularBufferSize,
+  );
+
+  const file = await fsp.open(filename, 'r+');
+
+  const headerBuffer = Buffer.alloc(I_ARENA_START);
+
+  await file.read({
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js:204]: should this check `bytesRead` ?

**Comment by mhofman** [comment]
> Thanks for pointing that out. I'm convinced I can make the writes async without any API changes. But it does add a bit of work so please let me know whether you think it's necessary.

I am slightly concerned by the perf impact of the sync fs API for something enabled by default by all validators. I would not be as concerned if the slog sender agent defaulted to a subprocess, which we should consider doing instead. If not, we do have utilities around to build async queues.

File: packages/telemetry/src/flight-recorder.js [hunk 9]
```diff
@@ -217,18 +181,186 @@ export const makeMemoryMappedCircularBuffer = async ({
...
+  arenaSize === hdrArenaSize ||
+    Fail`${filename} arena size mismatch; wanted ${arenaSize}, got ${hdrArenaSize}`;
+
+  /** @type {(outbuf: Uint8Array, readStart: number, firstReadLength: number) => void} */
+  const readRecord = (outbuf, readStart, firstReadLength) => {
+    fs.readSync(file.fd, outbuf, {
+      length: firstReadLength,
+      position: Number(readStart) + I_ARENA_START,
+    });
+    if (firstReadLength < outbuf.byteLength) {
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js]: I think it'd be preferable to get the number of `bytesRead` returned by `readSync`, and compare that to `firstReadLength`

File: packages/telemetry/src/flight-recorder.js [hunk 10]
```diff
@@ -217,18 +181,186 @@ export const makeMemoryMappedCircularBuffer = async ({
...
+      Number(circEnd) + header.byteLength,
+    );
+    if (firstWriteLength < record.byteLength) {
+      // Write to the beginning of the arena.
+      fs.writeSync(
+        file.fd,
+        record.subarray(firstWriteLength, record.byteLength),
+        undefined,
+        undefined,
+        header.byteLength,
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js]: I am confused by this position. Isn't it `I_ARENA_START` here?

File: packages/telemetry/src/flight-recorder.js [hunk 11]
```diff
@@ -217,18 +181,186 @@ export const makeMemoryMappedCircularBuffer = async ({
...
+   * @param {number} firstWriteLength
+   * @param {bigint} circEnd
+   */
+  const writeRecord = (record, firstWriteLength, circEnd) => {
+    fs.writeSync(
+      file.fd,
+      record.subarray(0, firstWriteLength),
+      // TS saying options bag not available
+      undefined,
+      undefined,
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js]: I believe it's more efficient to use offset + length

File: packages/telemetry/src/flight-recorder.js [hunk 12]
```diff
@@ -387,6 +387,6 @@ export const makeSlogSenderFromBuffer = ({ writeCircBuf }) => {
  * @type {import('./index.js').MakeSlogSender}
  */
 export const makeSlogSender = async opts => {
-  const { writeCircBuf } = await makeMemoryMappedCircularBuffer(opts);
+  const { writeCircBuf } = await makeSimpleCircularBuffer(opts);
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js]: Should it be an option for now if we're keeping both?

File: packages/telemetry/src/ingest-slog-entrypoint.js [hunk 13]
```diff
@@ -74,6 +74,7 @@ async function run() {
     if (!flush) {
       return;
     }
+    assert(slogSender.forceFlush, 'forceFlush missing from slogSender');
     await slogSender.forceFlush();
```
Comment (mhofman) [inline packages/telemetry/src/ingest-slog-entrypoint.js]: I think it'd be more correct to conditionally call.

```suggestion
    await slogSender.forceFlush?.();
```

File: packages/telemetry/src/flight-recorder.js [hunk 14]
```diff
@@ -371,18 +366,23 @@ export const makeMemoryMappedCircularBuffer = async ({
...
  */
 export const makeSlogSenderFromBuffer = ({ writeCircBuf }) => {
+  let writes = [];
   const writeJSON = obj => {
     const jsonObj = serializeSlogObj(obj);
     // Prepend a newline so that the file can be more easily manipulated.
     const data = new TextEncoder().encode(`\n${jsonObj}`);
     // console.log('have obj', obj, data);
-    writeCircBuf(data);
+    writes.push(writeCircBuf(data));
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js]: Since each write will update the arena you must make the writes sequential, aka chain the write to the completion of the previous promise. As such you don't keep an array of promise, but update single binding holding the latest promise.

**Review by mhofman — CHANGES_REQUESTED** [review]
I did a first pass reviewing commit-by-commit, spread over the day when I had time. As such some earlier comments may no longer be accurate.

The request for change is for the queueing mechanism, and the params of the `writeJSON`.

**Comment by mhofman** [comment]
> I think the existing test coverage is enough.

For good measure I'll force integration as that causes more stress on the flight-recorder

File: packages/telemetry/src/make-slog-sender.js (hunk 4 above)
Comment (turadg) [reply packages/telemetry/src/make-slog-sender.js]: probably a refactoring remnant. removed.

File: packages/telemetry/src/flight-recorder.js (hunk 5 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js]: Ah, I see. thanks. I've added a test.

File: packages/telemetry/src/flight-recorder.js (hunk 6 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js:15]: How so? This is part of 53588c8428dde829ddd703b9af0d01e56565e250 entitled "chore: remove mmap-based CircularBuffer"

File: packages/telemetry/src/flight-recorder.js
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js:1]: Aren't they? I suppose this is an artifact of my pushes while you were reviewing by commit. I'm pushing a clean rebase to make the proposed commit history apparent.

File: packages/telemetry/src/flight-recorder.js (hunk 7 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js:25]: Ah, more reason for this PR I suppose. Worth writing it there now?

File: packages/telemetry/src/flight-recorder.js (hunk 12 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js]: I assumed I was the only one interested in keeping both so once I had confidence in the new one I removed the old one.

Let me know if you think the mmap one should stick around until a later day

File: packages/telemetry/src/ingest-slog-entrypoint.js (hunk 13 above)
Comment (turadg) [reply packages/telemetry/src/ingest-slog-entrypoint.js]: thanks. at the point I wrote this it seemed like `forceFlush` wasn't truly optional but I see now it is.

File: packages/telemetry/src/flight-recorder.js (hunk 14 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js]: of course, thanks

File: packages/telemetry/src/flight-recorder.js (hunk 11 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js]: You mean instead of `subarray`? makes sense. incoming.

File: packages/telemetry/src/flight-recorder.js (hunk 10 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js]: they're equivalent  since the arena starts at the end of the header.

I've changed ot to `I_ARENA_START` since apparently that's more clear

File: packages/telemetry/src/flight-recorder.js
Comment (mhofman) [reply packages/telemetry/src/flight-recorder.js:1]: This was meant as a comment for commit `feat: simple CircularBuffer with fs offsets` (currently 58ed0b294d1387445e6d36bf53615eae87ea3a85)

It moves the BufferFromFile around and adds the new implementation in the same commit, which makes it hard to follow what is new and what is refactored

File: packages/telemetry/src/flight-recorder.js (hunk 6 above)
Comment (mhofman) [reply packages/telemetry/src/flight-recorder.js:15]: Yeah GH is bad at showing where comments were made.

The removal of the BufferFromFile import currently appears first in `refactor: fsp` (commit d092aec0f969af2f6af9d203f619edcb36263dab), and is reverted in `
feat: simple CircularBuffer with fs offsets` (58ed0b294d1387445e6d36bf53615eae87ea3a85), then removed again in `chore: remove mmap-based CircularBuffer` (c79b83eda85d8ebfbb07ec28d8d544775206e978).

The initial removal / re-introduction seems like a mistake in rebase / staging of partial changes.

File: packages/telemetry/src/flight-recorder.js [hunk 15]
```diff
@@ -272,18 +272,18 @@ export const makeSimpleCircularBuffer = async ({
...
+    toWrite = toWrite.then(() => writeCircBuf(data));
   };
   return Object.assign(writeJSON, {
     forceFlush: async () => {
-      const flushing = [...writes];
-      writes = [];
-      return Promise.all(flushing).then(() => {});
+      const writing = toWrite;
+      toWrite = Promise.resolve();
+      await writing;
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js]: Any reason not to simply do
```suggestion
      await toWrite;
```

File: packages/telemetry/src/flight-recorder.js [hunk 16]
```diff
@@ -272,18 +272,18 @@ export const makeSimpleCircularBuffer = async ({
...
  */
 export const makeSlogSenderFromBuffer = ({ writeCircBuf }) => {
-  let writes = [];
+  let toWrite = Promise.resolve();
   const writeJSON = (obj, serialized = serializeSlogObj(obj)) => {
     // Prepend a newline so that the file can be more easily manipulated.
     const data = new TextEncoder().encode(`\n${serialized}`);
     // console.log('have obj', obj, data);
-    writes.push(writeCircBuf(data));
+    toWrite = toWrite.then(() => writeCircBuf(data));
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js:283]: Just in case we start returning stuff from `writeCircBuf`, no need to hold onto it.

```suggestion
    toWrite = toWrite.then(() => { writeCircBuf(data); });
```

File: packages/telemetry/src/flight-recorder.js (hunk 9 above)
Comment (mhofman) [reply packages/telemetry/src/flight-recorder.js]: What do you think about this suggestion?

File: packages/telemetry/src/flight-recorder.js (hunk 12 above)
Comment (mhofman) [reply packages/telemetry/src/flight-recorder.js]: I'm fine removing. When I first started to review, it seemed it was sticking around.

**Review by mhofman — COMMENTED** [review]
Looks great so far.

I'd like to re-do a full pass after a rebase, especially if you can split the commit titled `
feat: simple CircularBuffer with fs offsets`

File: packages/telemetry/src/flight-recorder.js (hunk 15 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js]: I thought it would retain the promises forming the chain of thens. I take it that's not the case. I'll simplify.

File: packages/telemetry/src/flight-recorder.js (hunk 6 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js:15]: ah, yeah I read this in the full PR view. I'll amend

File: packages/telemetry/src/flight-recorder.js (hunk 9 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js:227]: no opinion. including.

File: packages/telemetry/src/flight-recorder.js (hunk 16 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js:283]: `writeCircBuf` has to be awaited. So this would be,
```js
toWrite = toWrite.then(async () => {
  await writeCircBuf(data);
});
```
which is pretty clunky. The type hints are all `Promise<void>` so nothing should rely on a return value.

File: packages/telemetry/src/flight-recorder.js (hunk 9 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js]: I read it as,
```js
    if (firstReadLength < bytesRead) {
```
but that doesn't make sense to me because they'll always equal. I suppose you meant,
```js
    if (bytesRead < outbuf.length) {
```

that reads well to me. "If bytes read are less than the buffer length". changed.

File: packages/telemetry/src/flight-recorder.js (hunk 16 above)
Comment (mhofman) [reply packages/telemetry/src/flight-recorder.js:283]: Oh my bad! Yeah ok too clunky

File: packages/telemetry/src/flight-recorder.js (hunk 9 above)
Comment (mhofman) [reply packages/telemetry/src/flight-recorder.js]: I guess I'm confused as to what `firstReadLength` holds. I should go back to the logic.

My concern was mostly that I believe `fs.read` can return less bytes than what asked if it reaches EOF.

File: packages/telemetry/src/flight-recorder.js [hunk 17]
```diff
@@ -218,34 +178,115 @@ export const makeMemoryMappedCircularBuffer = async ({
...
-    if (firstWriteLength < record.byteLength) {
-      // Write to the beginning of the arena.
-      arena.set(record.subarray(firstWriteLength, record.byteLength), 0);
-    }
     header.setBigUint64(
       I_CIRC_END,
       (circEnd + BigInt(record.byteLength)) % arenaSize,
     );
+
+    return writeRecord(record, firstWriteLength, circEnd);
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js:175]: I note that technically the arena header is now updated before the record is written, but that is not consequential

File: packages/telemetry/src/flight-recorder.js [hunk 18]
```diff
@@ -165,14 +122,17 @@ export const makeMemoryMappedCircularBuffer = async ({
...
-    if (RECORD_HEADER_SIZE + data.byteLength > arena.byteLength) {
+    if (RECORD_HEADER_SIZE + data.byteLength > arenaSize) {
       // Silently drop, it just doesn't fit.
       return;
     }
 
+    // Allocate for the data and a header
     const record = new Uint8Array(RECORD_HEADER_SIZE + data.byteLength);
+    // Set the data, after the header
     record.set(data, RECORD_HEADER_SIZE);
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js:122]: I never realized we were copying the data so many times, but that under the future improvement category.

File: packages/telemetry/src/flight-recorder.js [hunk 19]
```diff
@@ -189,6 +191,98 @@ function finishCircularBuffer(arenaSize, header, readRecord, writeRecord) {
...
+    Fail`${filename} arena size mismatch; wanted ${arenaSize}, got ${hdrArenaSize}`;
+
+  /** @type {(outbuf: Uint8Array, readStart: number, firstReadLength: number) => void} */
+  const readRecord = (outbuf, readStart, firstReadLength) => {
+    const bytesRead = fs.readSync(file.fd, outbuf, {
+      length: firstReadLength,
+      position: Number(readStart) + I_ARENA_START,
+    });
+
+    if (bytesRead < outbuf.byteLength) {
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js]: Ok so I went back to read the caller's logic, and basically we can use either `bytesRead` or `firstReadLength` here because we should technically be checking that `bytesRead === firstReadLength` in the first place.

Since we're pre-allocating a file the data should be available, but there are technically signal cases where the [read may still return less data than was requested](https://man7.org/linux/man-pages/man2/read.2.html#RETURN_VALUE). 

None of the read in this implementation seem to check the actual number of bytes read.

File: packages/telemetry/src/flight-recorder.js [hunk 20]
```diff
@@ -189,6 +191,98 @@ function finishCircularBuffer(arenaSize, header, readRecord, writeRecord) {
...
+   * @param {number} firstWriteLength
+   * @param {bigint} circEnd
+   */
+  const writeRecord = async (record, firstWriteLength, circEnd) => {
+    await file.write(
+      record,
+      // TS saying options bag not available
+      0,
+      firstWriteLength,
+      Number(circEnd) + header.byteLength,
```
Comment (mhofman) [inline packages/telemetry/src/flight-recorder.js]: Nit: to be consistent this should also be `I_ARENA_START` (or revert the others and use `header.byteLength` everywhere)

**Review by mhofman — APPROVED** [review]
Thanks for refactoring this!

Approving, but I removed the automerge label in case you want to address the read bytes / length checks.

File: packages/telemetry/src/flight-recorder.js (hunk 19 above)
Comment (turadg) [reply packages/telemetry/src/flight-recorder.js]: Thanks for pointing out bytesRead could come up short. This never had a way to handle that so I'll just assert that it didn't happen.
